### PR TITLE
ENH: Artifact.validate and validation levels

### PR DESCRIPTION
**Pull request by thermokarst**

### CODE REVIEW

File: qiime2/sdk/result.py [hunk 1]
```diff
@@ -252,6 +252,16 @@ def has_metadata(self):
...
 
+    def validate(self, level='max'):
+        """ Validates the data contents of an artifact
+
+        Raises
+        ------
+        ValidationError
+            If the artifact is invalid at the specified level of validation.
+        """
+        self.format.validate(self.view(self.format), level)
```
Comment (ebolyen) [inline qiime2/sdk/result.py]: 👍

File: qiime2/plugin/model/base.py [hunk 2]
```diff
@@ -35,3 +35,10 @@ def __str__(self):
 
 class ValidationError(Exception):
     pass
+
+
+# TODO: once sniff is dropped, move this up into FormatBase as validate method
+def _check_validation_level(level):
+    if level not in ('min', 'max'):
+        raise ValueError('Invalid validation level requested (%s), must '
+                         'be min or max.' % level)
```
Comment (ebolyen) [inline qiime2/plugin/model/base.py]: `'min' or 'max'`

File: qiime2/core/testing/format.py [hunk 3]
```diff
@@ -13,24 +13,37 @@
 class IntSequenceFormat(TextFileFormat):
     """
     A sequence of integers stored on new lines in a file. Since this is a
-    sequence, the integers have an order and repetition of elements is allowed.
+    sequence, the integers have an order and repetition of elements is
+    allowed. Sequential values must have an inter-value distance other than 3
+    to be valid.
```
Comment (ebolyen) [inline qiime2/core/testing/format.py]: That's a pretty intense constraint!

File: qiime2/sdk/plugin_manager.py [hunk 4]
```diff
@@ -85,6 +85,15 @@ def _integrate_plugin(self, plugin):
                     " and %r" %
                     (name, record.plugin.name, self.formats[name].plugin.name)
                 )
+
+            fmt = record.format
```
Comment (ebolyen) [inline qiime2/sdk/plugin_manager.py]: Could you add a `#TODO` to remove when sniff is removed? The `Format` would use an ABC at that point.

**Comment by thermokarst** [comment]
Thanks @ebolyen! I believe I have [addressed](https://github.com/qiime2/qiime2/pull/347/commits/b5071c19fd506845ee8a8845fc440634b1926fbf) your comments.

**Pull request merged**